### PR TITLE
bugfix for `EventContent` renaming

### PR DESCRIPTION
**Pull request by missirol**
If the HLT configuration contains more than 1 EventContent, and one tries to rename any one of the EventContents, the GUI ~~crashes~~ throws an Exception (but remains partly functional) when one hits Enter after typing the new EventContent name.

I see that the crash occurs in
https://github.com/cms-sw/hlt-confdb/blob/70306794653596f6f0c93bcaab6580ccdf2e8ca0/src/confdb/gui/ConfigurationTreeEditor.java#L139
the second time `.next()` is called (I suspect this is why there is no crash if there is only 1 EventContent).

An example of the stack trace is below.

I have no understanding of this, but somehow the simple change in this PR seems to avoid the crash (plus I tried to rename EventContents, save, and re-open a couple of configs, and things look correct).

Any insight, or suggestions?

---

Example of stack trace:
```
Exception in thread "AWT-EventQueue-0" java.util.ConcurrentModificationException
	at java.base/java.util.ArrayList$Itr.checkForComodification(ArrayList.java:1043)
	at java.base/java.util.ArrayList$Itr.next(ArrayList.java:997)
	at confdb.gui.ConfigurationTreeEditor.getCellEditorValue(ConfigurationTreeEditor.java:152)
	at java.desktop/javax.swing.plaf.basic.BasicTreeUI.completeEditing(BasicTreeUI.java:2382)
	at java.desktop/javax.swing.plaf.basic.BasicTreeUI$Handler.editingStopped(BasicTreeUI.java:4117)
	at java.desktop/javax.swing.AbstractCellEditor.fireEditingStopped(AbstractCellEditor.java:147)
	at java.desktop/javax.swing.DefaultCellEditor$EditorDelegate.stopCellEditing(DefaultCellEditor.java:370)
	at java.desktop/javax.swing.DefaultCellEditor.stopCellEditing(DefaultCellEditor.java:234)
	at java.desktop/javax.swing.DefaultCellEditor$EditorDelegate.actionPerformed(DefaultCellEditor.java:387)
	at java.desktop/javax.swing.JTextField.fireActionPerformed(JTextField.java:508)
	at java.desktop/javax.swing.JTextField.postActionEvent(JTextField.java:723)
	at java.desktop/javax.swing.JTextField$NotifyAction.actionPerformed(JTextField.java:839)
	at java.desktop/javax.swing.SwingUtilities.notifyAction(SwingUtilities.java:1810)
	at java.desktop/javax.swing.JComponent.processKeyBinding(JComponent.java:2900)
	at java.desktop/javax.swing.JComponent.processKeyBindings(JComponent.java:2948)
	at java.desktop/javax.swing.JComponent.processKeyEvent(JComponent.java:2862)
	at java.desktop/java.awt.Component.processEvent(Component.java:6412)
	at java.desktop/java.awt.Container.processEvent(Container.java:2263)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5011)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2321)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4843)
	at java.desktop/java.awt.KeyboardFocusManager.redispatchEvent(KeyboardFocusManager.java:1950)
	at java.desktop/java.awt.DefaultKeyboardFocusManager.dispatchKeyEvent(DefaultKeyboardFocusManager.java:870)
	at java.desktop/java.awt.DefaultKeyboardFocusManager.preDispatchKeyEvent(DefaultKeyboardFocusManager.java:1139)
	at java.desktop/java.awt.DefaultKeyboardFocusManager.typeAheadAssertions(DefaultKeyboardFocusManager.java:1009)
	at java.desktop/java.awt.DefaultKeyboardFocusManager.dispatchEvent(DefaultKeyboardFocusManager.java:835)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:4892)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2321)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2772)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4843)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:772)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:745)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:743)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:742)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
```

### CODE REVIEW

**Comment by missirol** [comment]
Correction: the GUI does not crash, it throws an Exception and remains partly functional.

File: src/confdb/gui/ConfigurationTreeEditor.java [hunk 1]
```diff
@@ -134,9 +134,9 @@ public Object getCellEditorValue() {
 			EventContent content = (EventContent) toBeEdited;
 			content.setName(name);
 			treeModel.nodeChanged(content);
-			Iterator<EventContent> itC = config.contentIterator();
-			while (itC.hasNext())
-				treeModel.nodeChanged(itC.next());
+			for(int idx = 0; idx < config.contentCount(); ++idx){
+			    treeModel.nodeChanged(config.content(idx));
+			}
```
Comment (Martin-Grunewald) [inline src/confdb/gui/ConfigurationTreeEditor.java:154]: I actually wonder why all EventContents existing in the config need to have treeModel.nodeChanged called, as opposed to just the single one being edited (in line 136). Maybe one could actually get rid of the loop over all?

Comment (missirol) [reply src/confdb/gui/ConfigurationTreeEditor.java:154]: I tried, but the result is not as good graphically. Without the loop, the renamed EventContent will move to its new position (in alphabetical order), but its original slot will remain blank, and one of the other EventContents won't be shown anymore. The loop is somehow updating them all such that they are all displayed and in alphabetical order.

Comment (Martin-Grunewald) [reply src/confdb/gui/ConfigurationTreeEditor.java:154]: Maybe get rid of line 136 only? It sets as nodeChanged the current content one and then loops over all contents, so the current content one gets nodeChanged twice?

Comment (missirol) [reply src/confdb/gui/ConfigurationTreeEditor.java:154]: Tried, but it doesn't do the trick.. without this PR, the GUI seems to behave the same with or without L136. Trying to have a closer look.

**Comment by missirol** [comment]
Some notes taken while debugging.

  - I checked that the crash comes from `.nodeChanged(itC.next())` (regardless of what is done before that call).

  - I tried to look up the exception message
    ```
    Exception in thread "AWT-EventQueue-0" java.util.ConcurrentModificationException
    ```
    It sounds like the issue is that we are iterating over a collection, while trying to edit the object that we are accessing via the iterator. [This short article](https://foystor.github.io/2022/06/02/ConcurrentModificationException/) mentions that one solution is to use a `for` loop as in this PR.

Trying to isolate the crash.

 - The crash happens in this call to `nodeChanged`:
https://github.com/cms-sw/hlt-confdb/blob/6dd9ea2474d9b3b7cc031d9af51858b45829dabc/src/confdb/gui/ConfigurationTreeEditor.java#L139

 - Inside `nodeChanged`, the crash occurs if `getIndexOfChild` is called (if one comments `getIndexOfChild`, the crash does not occur):
https://github.com/cms-sw/hlt-confdb/blob/6dd9ea2474d9b3b7cc031d9af51858b45829dabc/src/confdb/gui/tree/AbstractTreeModel.java#L56

 - In this case, `getIndexOfChild` leads to a call to `(Configuration).indexOfContent`:
https://github.com/cms-sw/hlt-confdb/blob/6dd9ea2474d9b3b7cc031d9af51858b45829dabc/src/confdb/gui/ConfigurationTreeModel.java#L685

 - `(Configuration).indexOfContent` has the peculiarity of modifying the `contents` member of the `Configuration` object, because `contents` gets sorted at every call:
https://github.com/cms-sw/hlt-confdb/blob/6dd9ea2474d9b3b7cc031d9af51858b45829dabc/src/confdb/data/Configuration.java#L2169

 - If one simply comments the sorting with the patch in [1], the crash does not occur (but then the GUI won't re-order the EventContents). Basically (and this is not an exact explanation), it looks like `nodeChanged` leads to modifying `(Configuration).contents` because of sorting, and this clashes with the use of iterators (i.e. `next()`) in `ConfigurationTreeEditor`.

 - Using a simple `for` loop in `ConfigurationTreeEditor` avoids this issue. Note that the [first call to `nodeChanged` in `ConfigurationTreeEditor`](https://github.com/cms-sw/hlt-confdb/blob/6dd9ea2474d9b3b7cc031d9af51858b45829dabc/src/confdb/gui/ConfigurationTreeEditor.java#L136) follows the actual renaming of the EventContent in question (see `setName`), and this is where the EventContents are re-sorted for the first time, and their order can change; [the second call to `nodeChanged`](https://github.com/cms-sw/hlt-confdb/blob/6dd9ea2474d9b3b7cc031d9af51858b45829dabc/src/confdb/gui/ConfigurationTreeEditor.java#L139) in something like
```java
for(int idx = 0; idx < config.contentCount(); ++idx){
  treeModel.nodeChanged(config.content(idx));
}
```
is needed to display all the EventContents in their new order. The `(Configuration).contents` get sorted at every call to `nodeChanged`, but their order does not change (because their names are set), so a simple `for` loop does access all the EventContents.

This is my understanding, but there might be things I'm missing.

[1]
```diff
diff --git a/src/confdb/data/Configuration.java b/src/confdb/data/Configuration.java
index b10a80d7..6efcc7b5 100644
--- a/src/confdb/data/Configuration.java
+++ b/src/confdb/data/Configuration.java
@@ -2152,7 +2152,7 @@ public class Configuration implements IConfiguration {
 
        /** retrieve i-th event content */
        public EventContent content(int i) {
-               Collections.sort(contents);
+//             Collections.sort(contents);
                return contents.get(i);
        }
 
@@ -2166,13 +2166,13 @@ public class Configuration implements IConfiguration {
 
        /** index of a certain event content */
        public int indexOfContent(EventContent ec) {
-               Collections.sort(contents);
+//             Collections.sort(contents);
                return contents.indexOf(ec);
        }
 
        /** retrieve event content iterator */
        public Iterator<EventContent> contentIterator() {
-               Collections.sort(contents);
+//             Collections.sort(contents);
                return contents.iterator();
        }
```

**Comment by missirol** [comment]
I made a couple of small updates on top of the initial version of the PR, based on the debugging described above.

I have tested the PR locally (editing configs, saving them, and re-opening them), and it works as expected.

I'm not saying I fully understand this, but at this point I vote for integrating the PR as is.

**Comment by missirol** [comment]
@Martin-Grunewald @Sam-Harper 

Could you please let me know if you have feedback on this PR ? I think it could be merged.

**Comment by Martin-Grunewald** [comment]
Yes, go ahead!

**Comment by missirol** [comment]
Okay, thanks!

After this merge, I plan to make a new 'minor release' version of the GUI by next Monday. (The reason for the release is that I don't know when I can work on some of the other updates, and some might require changes to the python output, which implies a new 'major' release).

**Pull request merged**